File: Problemset/min-stack/min-stack.go

// @Title: 最小栈 (Min Stack)
// @Author: KivenC
// @Date: 2019-07-10 15:21:23
// @Runtime: 32 ms
// @Memory: 7.4 MB

type MinStack struct {
	Stack  []int
	Helper []int
}

/** initialize your data structure here. */
func Constructor() MinStack {
	return MinStack{[]int{}, []int{}}
}

func (this *MinStack) Push(x int) {
	this.Stack = append(this.Stack, x)
	if len(this.Helper) > 0 && this.Helper[len(this.Helper)-1] < x {
		this.Helper = append(this.Helper, this.Helper[len(this.Helper)-1])
    } else {
        this.Helper = append(this.Helper, x)
    }
}

func (this *MinStack) Pop() {
	this.Helper = this.Helper[:len(this.Helper)-1]
	this.Stack = this.Stack[:len(this.Stack)-1]
}

func (this *MinStack) Top() int {
	return this.Stack[len(this.Stack)-1]
}

func (this *MinStack) GetMin() int {
	return this.Helper[len(this.Helper)-1]
}



/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */

File: Problemset/min-stack/min-stack.py

# @Title: 最小栈 (Min Stack)
# @Author: KivenC
# @Date: 2019-07-10 14:52:28
# @Runtime: 72 ms
# @Memory: 16.1 MB

class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.helper = []  # 辅助栈

    def push(self, x: int) -> None:
        self.stack.append(x)
        if self.helper and self.helper[-1] < x:
            self.helper.append(self.helper[-1])
        else:
            self.helper.append(x)

    def pop(self) -> None:
        self.stack.pop()
        self.helper.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.helper[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
